src/posts/2024-02-12-how-software-networking-is-taking-over-the-world/index.md

```diff
@@ -0,0 +1,63 @@
+
+## Software Catches Up
+
+In the years since that paper was written, there has been a huge evolution in the commercial hardware, CPUs have grown horizontally instead of vertically (single CPU core speed growth has all but stopped, and the focus of chip manufacturers has been multiple packaged cores).
+
+- 2015: Intel Core i7-6700 3.4Ghz (4 cores, 4 threads)
+- 2017: Intel Core i7-8700 3.2Ghz (6 cores, 12 threads)
+- 2019: Intel Core i7-9700 3.0Ghz (8 cores, 8 threads)
+- 2021: Intel Core i7-11700 2.5Ghz (8 cores, 16 threads)
+- 2023: Intel Core i7-13700 2.1Ghz (16 cores, 24 threads)
+
+\*_data taken from [eatyourbytes.com](https://www.eatyourbytes.com/launchyear/2023/) (and subsequent years)_
+
+It would be a huge disservice to point out that the efficiency of single-core processing also hasn’t hugely increased, however the above is to illustrate that the architecture is changing. The network cards themselves sit on much faster connections to the CPU itself such as PCI-Express, and often come with offloading engines for certain repetitive tasks reducing the load on the CPU (such as a TCP offloading engine).
+
+Additionally, the application architecture has had a huge shift, due to the increase in processing capacity the model typically runs multiple workloads on a single machine either through virtualization or containerisation. This has dramatically increased the amount of east-west traffic both on a single network and actually within a server itself. This leads to a requirement for processing network packets within the server itself; having to push traffic out of the server to an external network device for network reasons here is going to increase the latency for the application.
+
+As mentioned above both the hardware **and** the evolution of the application architecture has pushed the requirement for more localized network processing. However, until relatively recently, this would still be a very inefficient model because processing packets in userland comes with huge performance penalties. For each packet, the kernel will first need to perform its workflow and then it will need to copy that packet to userland, where lower-priority application code can process it and finally return a response, leading to another data copy as the packet transitions into the kernel. For higher performance, we would need to write code that runs inside the kernel where it is largely free of the overhead of user programs etc. However pushing code upstream to the Linux kernel has to go through a rigorous process, and there is no guarantee that it would be accepted. You could potentially manage your copy of a kernel with your network code in and accept the huge amount of technical debt moving forward. Kernel modules are a potential option, but they still come with the overhead and technical debt of managing a continuing stream of new releases and patching for older kernels, not to mention the risk of bugs bringing down the whole server.
```

Comment:
```suggestion
As mentioned above both the hardware **and** the evolution of the application architecture has pushed the requirement for more localized network processing. However, until relatively recently, this would still be a very inefficient model because processing packets in userland comes with huge performance penalties. For each packet, the kernel will first need to perform its workflow and then it will need to copy that packet to userland, where lower-priority application code can process it and finally return a response, leading to another data copy as the packet transitions into the kernel. For higher performance, we would need to write code that runs inside the kernel where it is largely free of the overhead of user programs etc. However pushing code upstream to the Linux kernel has to go through a rigorous process, and there is no guarantee that it would be accepted.

You could potentially manage a copy of a kernel with your network code in and accept a huge amount of technical debt moving forward. Kernel modules are also a potential option, but they still come with the overhead and technical debt of managing a continuing stream of new releases and patching for older kernels, not to mention the risk of bugs bringing down the whole server.
```